Confusing error when checking trait impl against trait

When I try to check the following trait implementation ([here](https://github.com/PLSysSec/tock/blob/793c017ef05edb13c51215c7bef0622dd1735cb3/kernel/src/collections/ring_buffer.rs#L178-L184)):
```Rust
    #[flux_rs::sig(
        fn(self: &strg Self, _) -> Option<_> ensures self: Self
    )]
    fn remove_first_matching<F>(&mut self, f: F) -> Option<T>
    where
        F: Fn(&T) -> bool,
    {...}
```
against the trait spec ([here](https://github.com/PLSysSec/tock/blob/793c017ef05edb13c51215c7bef0622dd1735cb3/kernel/src/collections/queue.rs#L30-L35)):
```Rust
    #[flux_rs::sig(
        fn(self: &strg Self, _) -> Option<_> ensures self: Self
    )]
    fn remove_first_matching<F>(&mut self, f: F) -> Option<T>
    where
        F: Fn(&T) -> bool;
```
it fails with the error:
```
error[E0999]: internal flux error: crates/flux-infer/src/infer.rs:533:22
   --> kernel/src/collections/ring_buffer.rs:181:5
    |
181 | /     fn remove_first_matching<F>(&mut self, f: F) -> Option<T>
182 | |     where
183 | |         F: Fn(&T) -> bool,
    | |__________________________^
    |
    = note: incompatible types: `&'_ strg <self: RingBuffer<'_, λb0. T[b0]>[RingBuffer { a0, a1, a2 }]>` - `&'_ strg <?0e#8242: ∃b0. RingBuffer<'_, λb1. T[b1]>[b0]>`
```

I'm a little confused because the spec is token-for-token identical in both the trait and implementation, and only asserts that we need a `&strg Self`. 

I also get a similar error when there is no spec on the trait at all. I don't fully understand the implementation for trait_impl checking, so I'm not sure if this is expected or not.

 Any guidance?

--- a/crates/flux-infer/src/infer.rs
+++ b/crates/flux-infer/src/infer.rs
@@ -455,6 +455,20 @@ struct Sub {
     obligations: Vec<rty::Clause>,
 }
 
+/// [NOTE:unfold_strg_ref] We use this function to unfold a strong reference prior to a subtyping check.
+/// Normally, when checking a function body, a `StrgRef` is automatically unfolded
+/// i.e. `x:&strg T` is turned into turned into a `x:Ptr(l); l: T` where `l` is some
+/// fresh location. However, we need the below to do a similar unfolding in `check_fn_subtyping`
+/// where we just have the super-type signature that needs to be unfolded.
+/// We also add the binding to the environment so that we can:
+/// (1) UPDATE the location after the call, and
+/// (2) CHECK the relevant `ensures` clauses of the super-sig.
+/// Nico: More importantly, we are assuming functions always give back the "ownership"
+/// of the location so even though we should technically "consume" the ownership and
+/// remove the location from the environment, the type is always going to be overwritten.
+/// (there's a check for this btw, if you write an &strg we require an ensures for that
+/// location for the signature to be well-formed)
+
 impl Sub {
     fn new(reason: ConstrReason, span: Span) -> Self {
         Self { reason, span, obligations: vec![] }
@@ -487,7 +501,7 @@ impl Sub {
                 self.tys(infcx, &ty1, ty2)
             }
             (TyKind::StrgRef(_, path1, ty1), TyKind::StrgRef(_, path2, ty2)) => {
-                env.unfold_strg_ref(infcx, path1, ty1)?;
+                env.unfold_strg_ref(infcx, path1, ty1)?; // see [NOTE:unfold_strg_ref]
                 let ty1 = env.get(path1);
                 infcx.unify_exprs(&path1.to_expr(), &path2.to_expr());
                 self.tys(infcx, &ty1, ty2)

--- a/crates/flux-infer/src/infer.rs
+++ b/crates/flux-infer/src/infer.rs
@@ -48,6 +48,14 @@ impl Tag {
 }
 
 #[derive(PartialEq, Eq, Clone, Copy, Hash, Debug)]
+pub enum SubtypeReason {
+    Input,
+    Output,
+    Requires,
+    Ensures,
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Debug)]
 pub enum ConstrReason {
     Call,
     Assign,
@@ -58,11 +66,8 @@ pub enum ConstrReason {
     Rem,
     Goto(BasicBlock),
     Overflow,
+    Subtype(SubtypeReason),
     Other,
-    SubtypeIn,
-    SubtypeReq,
-    SubtypeOut,
-    SubtypeEns,
 }
 
 pub struct InferCtxtRoot<'genv, 'tcx> {

--- a/crates/flux-infer/src/infer.rs
+++ b/crates/flux-infer/src/infer.rs
@@ -59,6 +59,10 @@ pub enum ConstrReason {
     Goto(BasicBlock),
     Overflow,
     Other,
+    SubtypeIn,
+    SubtypeReq,
+    SubtypeOut,
+    SubtypeEns,
 }
 
 pub struct InferCtxtRoot<'genv, 'tcx> {

--- a/crates/flux-infer/src/infer.rs
+++ b/crates/flux-infer/src/infer.rs
@@ -11,7 +11,8 @@ use flux_middle::{
         fold::TypeFoldable,
         AliasKind, AliasTy, BaseTy, Binder, BoundVariableKinds, CoroutineObligPredicate, ESpan,
         EVar, EVarGen, EarlyBinder, Expr, ExprKind, GenericArg, GenericArgs, HoleKind, InferMode,
-        Lambda, List, Mutability, Path, PolyVariant, PtrKind, Ref, Region, Sort, Ty, TyKind, Var,
+        Lambda, List, Loc, Mutability, Path, PolyVariant, PtrKind, Ref, Region, Sort, Ty, TyKind,
+        Var,
     },
 };
 use itertools::{izip, Itertools};
@@ -431,6 +432,7 @@ pub trait LocEnv {
         bound: Ty,
     ) -> InferResult<Ty>;
 
+    fn unfold_strg_ref(&mut self, infcx: &mut InferCtxt, path: &Path, ty: &Ty) -> InferResult<Loc>;
     fn get(&self, path: &Path) -> Ty;
 }
 
@@ -471,6 +473,12 @@ impl Sub {
                 self.fun_args(infcx, env, a, ty_b)
             }
             (TyKind::Ptr(PtrKind::Mut(_), path1), TyKind::StrgRef(_, path2, ty2)) => {
+                let ty1 = env.get(path1);
+                infcx.unify_exprs(&path1.to_expr(), &path2.to_expr());
+                self.tys(infcx, &ty1, ty2)
+            }
+            (TyKind::StrgRef(_, path1, ty1), TyKind::StrgRef(_, path2, ty2)) => {
+                env.unfold_strg_ref(infcx, path1, ty1)?;
                 let ty1 = env.get(path1);
                 infcx.unify_exprs(&path1.to_expr(), &path2.to_expr());
                 self.tys(infcx, &ty1, ty2)
@@ -530,7 +538,7 @@ impl Sub {
                 }
                 Ok(())
             }
-            _ => Err(query_bug!("incompatible types: `{a:?}` - `{b:?}`"))?,
+            _ => panic!("incompatible types: `{a:?}` - `{b:?}`"), // _ => Err(query_bug!("incompatible types: `{a:?}` - `{b:?}`"))?,
         }
     }
 

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -235,7 +235,7 @@ fn find_trait_item(
 ///  fn g(x1:T1,...,xn:Tn) -> T {
 ///      f(x1,...,xn)
 ///  }
-/// TODO: copy rules from SLACK.
+#[expect(clippy::too_many_arguments)]
 fn check_fn_subtyping(
     infcx: &mut InferCtxt,
     def_id: &DefId,

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -4,7 +4,7 @@ use flux_common::{bug, dbg, index::IndexVec, iter::IterExt, tracked_span_bug};
 use flux_config as config;
 use flux_infer::{
     fixpoint_encoding::{self, KVarGen},
-    infer::{ConstrReason, InferCtxt, InferCtxtAt, InferCtxtRoot, SubtypeReason},
+    infer::{ConstrReason, InferCtxt, InferCtxtRoot, SubtypeReason},
     refine_tree::{AssumeInvariants, RefineTree, Snapshot},
 };
 use flux_middle::{
@@ -13,10 +13,9 @@ use flux_middle::{
     query_bug,
     rty::{
         self, fold::TypeFoldable, refining::Refiner, AdtDef, BaseTy, Binder, Bool, Clause,
-        CoroutineObligPredicate, EarlyBinder, Ensures, Expr, FnOutput, FnTraitPredicate,
-        GenericArg, GenericArgs, GenericArgsExt as _, Int, IntTy, Mutability, Path, PolyFnSig,
-        PtrKind, Ref, RefineArgs, RefineArgsExt, Region::ReStatic, Ty, TyKind, Uint, UintTy,
-        VariantIdx,
+        CoroutineObligPredicate, EarlyBinder, Expr, FnOutput, FnTraitPredicate, GenericArg,
+        GenericArgs, GenericArgsExt as _, Int, IntTy, Mutability, Path, PolyFnSig, PtrKind, Ref,
+        RefineArgs, RefineArgsExt, Region::ReStatic, Ty, TyKind, Uint, UintTy, VariantIdx,
     },
 };
 use flux_rustc_bridge::{
@@ -318,58 +317,14 @@ fn check_fn_subtyping(
         .with_span(span)?;
 
     // 6. Update state with Output "ensures" and check super ensures
-    update_ensures(&mut infcx, &mut env, &output, overflow_checking);
+    env.update_ensures(&mut infcx, &output, overflow_checking);
     fold_local_ptrs(&mut infcx, &mut env, span)?;
-    check_ensures(
-        &mut infcx,
-        &mut env,
-        &super_output,
-        ConstrReason::Subtype(SubtypeReason::Ensures),
-        span,
-    )?;
+    env.check_ensures(&mut infcx, &super_output, ConstrReason::Subtype(SubtypeReason::Ensures))
+        .with_span(span)?;
 
     let evars_sol = infcx.pop_scope().with_span(span)?;
     infcx.replace_evars(&evars_sol);
 
-    Ok(())
-}
-
-fn update_ensures(
-    infcx: &mut InferCtxt,
-    env: &mut TypeEnv,
-    output: &FnOutput,
-    overflow_checking: bool,
-) {
-    for ensure in &output.ensures {
-        match ensure {
-            Ensures::Type(path, updated_ty) => {
-                let updated_ty = infcx.unpack(updated_ty);
-                infcx.assume_invariants(&updated_ty, overflow_checking);
-                env.update_path(path, updated_ty);
-            }
-            Ensures::Pred(e) => infcx.assume_pred(e),
-        }
-    }
-}
-
-fn check_ensures(
-    at: &mut InferCtxtAt,
-    env: &mut TypeEnv,
-    output: &FnOutput,
-    reason: ConstrReason,
-    span: Span,
-) -> Result {
-    for constraint in &output.ensures {
-        match constraint {
-            Ensures::Type(path, ty) => {
-                let actual_ty = env.get(path);
-                at.subtyping(&actual_ty, ty, reason).with_span(span)?;
-            }
-            Ensures::Pred(e) => {
-                at.check_pred(e, ConstrReason::Ret);
-            }
-        }
-    }
     Ok(())
 }
 
@@ -779,7 +734,8 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             .subtyping(&ret_place_ty, &output.ret, ConstrReason::Ret)
             .with_span(span)?;
 
-        check_ensures(&mut at, env, &output, ConstrReason::Ret, span)?;
+        env.check_ensures(&mut at, &output, ConstrReason::Ret)
+            .with_span(span)?;
 
         let evars_sol = infcx.pop_scope().with_span(span)?;
         infcx.replace_evars(&evars_sol);
@@ -866,7 +822,7 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             .replace_evars(&evars_sol)
             .replace_bound_refts_with(|sort, _, _| infcx.define_vars(sort));
 
-        update_ensures(infcx, env, &output, self.check_overflow());
+        env.update_ensures(infcx, &output, self.check_overflow());
 
         fold_local_ptrs(infcx, env, span)?;
 

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -4,7 +4,7 @@ use flux_common::{bug, dbg, index::IndexVec, iter::IterExt, span_bug, tracked_sp
 use flux_config as config;
 use flux_infer::{
     fixpoint_encoding::{self, KVarGen},
-    infer::{ConstrReason, InferCtxt, InferCtxtRoot},
+    infer::{ConstrReason, InferCtxt, InferCtxtAt, InferCtxtRoot},
     refine_tree::{AssumeInvariants, RefineTree, Snapshot},
 };
 use flux_middle::{
@@ -243,6 +243,7 @@ fn check_fn_subtyping(
     sub_args: &[GenericArg],
     super_sig: EarlyBinder<rty::PolyFnSig>,
     super_args: Option<(&GenericArgs, &rty::RefineArgs)>,
+    overflow_checking: bool,
     span: Span,
 ) -> Result {
     let mut infcx = infcx.branch();
@@ -279,6 +280,8 @@ fn check_fn_subtyping(
         .normalize_projections(infcx.genv, infcx.region_infcx, *def_id)
         .with_span(span)?;
 
+    let mut env = TypeEnv::empty();
+
     // 3. INPUT subtyping (g-input <: f-input)
     // TODO: Check requires predicates (?)
     // for requires in fn_def_sig.requires() {
@@ -288,19 +291,23 @@ fn check_fn_subtyping(
         span_bug!(span, "Not yet handled: requires predicates {def_id:?}");
     }
     for (actual, formal) in iter::zip(actuals, sub_sig.inputs()) {
-        let (formal, pred) = formal.unconstr();
-        infcx.check_pred(&pred, ConstrReason::Call);
-        // see: TODO(pack-closure)
-        match (actual.kind(), formal.kind()) {
-            (TyKind::Ptr(rty::PtrKind::Mut(_), _), _) => {
-                bug!("Not yet handled: FnDef subtyping with Ptr");
-            }
-            _ => {
-                infcx
-                    .subtyping(&actual, &formal, ConstrReason::Call)
-                    .with_span(span)?;
-            }
-        }
+        infcx
+            .fun_arg_subtyping(&mut env, &actual, formal, ConstrReason::Call)
+            .with_span(span)?;
+
+        // let (formal, pred) = formal.unconstr();
+        // infcx.check_pred(&pred, ConstrReason::Call);
+        // // see: TODO(pack-closure)
+        // match (actual.kind(), formal.kind()) {
+        //     (TyKind::Ptr(rty::PtrKind::Mut(_), _), _) => {
+        //         bug!("Not yet handled: FnDef subtyping with Ptr");
+        //     }
+        //     _ => {
+        //         infcx
+        //             .subtyping(&actual, &formal, ConstrReason::Call)
+        //             .with_span(span)?;
+        //     }
+        // }
     }
 
     // 4. Plug in the EVAR solution / replace evars
@@ -314,26 +321,66 @@ fn check_fn_subtyping(
     // 5. OUTPUT subtyping (f_out <: g_out)
     // RJ: new `at` to avoid borrowing errors...!
     infcx.push_scope();
-    let oblig_output = super_sig
+    let super_output = super_sig
         .output()
         .replace_bound_refts_with(|sort, mode, _| infcx.fresh_infer_var(sort, mode));
     infcx
-        .subtyping(&output.ret, &oblig_output.ret, ConstrReason::Ret)
+        .subtyping(&output.ret, &super_output.ret, ConstrReason::Ret)
         .with_span(span)?;
     let evars_sol = infcx.pop_scope().with_span(span)?;
     infcx.replace_evars(&evars_sol);
 
-    if !output.ensures.is_empty() || !oblig_output.ensures.is_empty() {
-        span_bug!(span, "Not yet handled: subtyping with ensures predicates {def_id:?}");
-    }
+    // println!("TRACE: check_fn_subtyping (0): {env:?}");
+    // // 6. Update state with Output "ensures"
+    // update_ensures(&mut infcx, &mut env, &output, overflow_checking)?;
+    // println!("TRACE: check_fn_subtyping (1): {env:?}");
+    // // 7. Check ensures predicates
+    // check_ensures(&mut infcx, &mut env, &super_output, span)?;
 
     Ok(())
 }
+
+fn update_ensures(
+    infcx: &mut InferCtxt,
+    env: &mut TypeEnv,
+    output: &FnOutput,
+    overflow_checking: bool,
+) -> Result {
+    for ensure in &output.ensures {
+        match ensure {
+            Ensures::Type(path, updated_ty) => {
+                let updated_ty = infcx.unpack(updated_ty);
+                infcx.assume_invariants(&updated_ty, overflow_checking);
+                env.update_path(path, updated_ty);
+            }
+            Ensures::Pred(e) => infcx.assume_pred(e),
+        }
+    }
+    Ok(())
+}
+
+fn check_ensures(at: &mut InferCtxtAt, env: &mut TypeEnv, output: &FnOutput, span: Span) -> Result {
+    for constraint in &output.ensures {
+        match constraint {
+            Ensures::Type(path, ty) => {
+                let actual_ty = env.get(path);
+                at.subtyping(&actual_ty, ty, ConstrReason::Ret)
+                    .with_span(span)?;
+            }
+            Ensures::Pred(e) => {
+                at.check_pred(e, ConstrReason::Ret);
+            }
+        }
+    }
+    Ok(())
+}
+
 /// Trait subtyping check, which makes sure that the type for an impl method (def_id)
 /// is a subtype of the corresponding trait method.
 pub fn trait_impl_subtyping(
     genv: GlobalEnv,
     def_id: LocalDefId,
+    overflow_checking: bool,
     span: Span,
 ) -> Result<Option<(RefineTree, KVarGen)>> {
     // Skip the check if this is not an impl method
@@ -367,6 +414,7 @@ pub fn trait_impl_subtyping(
             &impl_args,
             trait_fn_sig,
             Some((&trait_args, &trait_refine_args)),
+            overflow_checking,
             span,
         )?;
         Ok(())
@@ -733,18 +781,7 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             .subtyping(&ret_place_ty, &output.ret, ConstrReason::Ret)
             .with_span(span)?;
 
-        for constraint in &output.ensures {
-            match constraint {
-                Ensures::Type(path, ty) => {
-                    let actual_ty = env.get(path);
-                    at.subtyping(&actual_ty, ty, ConstrReason::Ret)
-                        .with_span(span)?;
-                }
-                Ensures::Pred(e) => {
-                    at.check_pred(e, ConstrReason::Ret);
-                }
-            }
-        }
+        check_ensures(&mut at, env, &output, span)?;
 
         let evars_sol = infcx.pop_scope().with_span(span)?;
         infcx.replace_evars(&evars_sol);
@@ -831,16 +868,17 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             .replace_evars(&evars_sol)
             .replace_bound_refts_with(|sort, _, _| infcx.define_vars(sort));
 
-        for ensures in &output.ensures {
-            match ensures {
-                Ensures::Type(path, updated_ty) => {
-                    let updated_ty = infcx.unpack(updated_ty);
-                    infcx.assume_invariants(&updated_ty, self.check_overflow());
-                    env.update_path(path, updated_ty);
-                }
-                Ensures::Pred(e) => infcx.assume_pred(e),
-            }
-        }
+        update_ensures(infcx, env, &output, self.check_overflow())?;
+        // for ensures in &output.ensures {
+        //     match ensures {
+        //         Ensures::Type(path, updated_ty) => {
+        //             let updated_ty = infcx.unpack(updated_ty);
+        //             infcx.assume_invariants(&updated_ty, self.check_overflow());
+        //             env.update_path(path, updated_ty);
+        //         }
+        //         Ensures::Pred(e) => infcx.assume_pred(e),
+        //     }
+        // }
 
         fold_local_ptrs(infcx, env, span)?;
 
@@ -890,7 +928,16 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
                 // See `tests/neg/surface/fndef00.rs`
                 let sub_sig = self.genv.fn_sig(def_id).with_span(span)?;
                 let oblig_sig = fn_trait_pred.fndef_poly_sig();
-                check_fn_subtyping(infcx, def_id, sub_sig, args, oblig_sig, None, span)?;
+                check_fn_subtyping(
+                    infcx,
+                    def_id,
+                    sub_sig,
+                    args,
+                    oblig_sig,
+                    None,
+                    self.check_overflow(),
+                    span,
+                )?;
             }
             _ => {
                 // TODO: When we allow refining closure/fn at the surface level, we would need to do some function subtyping here,
@@ -1369,6 +1416,7 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
                         args,
                         super_sig,
                         None,
+                        self.check_overflow(),
                         stmt_span,
                     )?;
                     to

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -269,6 +269,8 @@ fn check_fn_subtyping(
         .map(|ty| infcx.unpack(ty))
         .collect_vec();
 
+    let mut env = TypeEnv::empty();
+    let actuals = unfold_local_ptrs(&mut infcx, &mut env, span, &sub_sig, &actuals)?;
     let actuals = infer_under_mut_ref_hack(&mut infcx, &actuals[..], sub_sig.as_ref());
 
     // 2. Fresh names for `T_f` refine-params / Instantiate fn_def_sig and normalize it
@@ -279,8 +281,6 @@ fn check_fn_subtyping(
         .replace_bound_vars(|_| rty::ReErased, |sort, mode| infcx.fresh_infer_var(sort, mode))
         .normalize_projections(infcx.genv, infcx.region_infcx, *def_id)
         .with_span(span)?;
-
-    let mut env = TypeEnv::empty();
 
     // 3. INPUT subtyping (g-input <: f-input)
     for requires in super_sig.requires() {
@@ -316,6 +316,7 @@ fn check_fn_subtyping(
 
     // 6. Update state with Output "ensures" and check super ensures
     update_ensures(&mut infcx, &mut env, &output, overflow_checking)?;
+    fold_local_ptrs(&mut infcx, &mut env, span)?;
     check_ensures(&mut infcx, &mut env, &super_output, ConstrReason::SubtypeEns, span)?;
 
     let evars_sol = infcx.pop_scope().with_span(span)?;

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -327,15 +327,16 @@ fn check_fn_subtyping(
     infcx
         .subtyping(&output.ret, &super_output.ret, ConstrReason::Ret)
         .with_span(span)?;
-    let evars_sol = infcx.pop_scope().with_span(span)?;
-    infcx.replace_evars(&evars_sol);
 
     // println!("TRACE: check_fn_subtyping (0): {env:?}");
     // // 6. Update state with Output "ensures"
-    // update_ensures(&mut infcx, &mut env, &output, overflow_checking)?;
+    update_ensures(&mut infcx, &mut env, &output, overflow_checking)?;
     // println!("TRACE: check_fn_subtyping (1): {env:?}");
     // // 7. Check ensures predicates
-    // check_ensures(&mut infcx, &mut env, &super_output, span)?;
+    check_ensures(&mut infcx, &mut env, &super_output, span)?;
+
+    let evars_sol = infcx.pop_scope().with_span(span)?;
+    infcx.replace_evars(&evars_sol);
 
     Ok(())
 }
@@ -869,16 +870,6 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             .replace_bound_refts_with(|sort, _, _| infcx.define_vars(sort));
 
         update_ensures(infcx, env, &output, self.check_overflow())?;
-        // for ensures in &output.ensures {
-        //     match ensures {
-        //         Ensures::Type(path, updated_ty) => {
-        //             let updated_ty = infcx.unpack(updated_ty);
-        //             infcx.assume_invariants(&updated_ty, self.check_overflow());
-        //             env.update_path(path, updated_ty);
-        //         }
-        //         Ensures::Pred(e) => infcx.assume_pred(e),
-        //     }
-        // }
 
         fold_local_ptrs(infcx, env, span)?;
 

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -1,6 +1,6 @@
 use std::{collections::hash_map::Entry, iter};
 
-use flux_common::{bug, dbg, index::IndexVec, iter::IterExt, span_bug, tracked_span_bug};
+use flux_common::{bug, dbg, index::IndexVec, iter::IterExt, tracked_span_bug};
 use flux_config as config;
 use flux_infer::{
     fixpoint_encoding::{self, KVarGen},
@@ -283,31 +283,17 @@ fn check_fn_subtyping(
     let mut env = TypeEnv::empty();
 
     // 3. INPUT subtyping (g-input <: f-input)
-    // TODO: Check requires predicates (?)
-    // for requires in fn_def_sig.requires() {
-    //     at.check_pred(requires, ConstrReason::Call);
-    // }
-    if !sub_sig.requires().is_empty() {
-        span_bug!(span, "Not yet handled: requires predicates {def_id:?}");
+    for requires in super_sig.requires() {
+        infcx.assume_pred(requires);
     }
     for (actual, formal) in iter::zip(actuals, sub_sig.inputs()) {
         infcx
-            .fun_arg_subtyping(&mut env, &actual, formal, ConstrReason::Call)
+            .fun_arg_subtyping(&mut env, &actual, formal, ConstrReason::SubtypeIn)
             .with_span(span)?;
-
-        // let (formal, pred) = formal.unconstr();
-        // infcx.check_pred(&pred, ConstrReason::Call);
-        // // see: TODO(pack-closure)
-        // match (actual.kind(), formal.kind()) {
-        //     (TyKind::Ptr(rty::PtrKind::Mut(_), _), _) => {
-        //         bug!("Not yet handled: FnDef subtyping with Ptr");
-        //     }
-        //     _ => {
-        //         infcx
-        //             .subtyping(&actual, &formal, ConstrReason::Call)
-        //             .with_span(span)?;
-        //     }
-        // }
+    }
+    // we check the requires AFTER the actual-formal subtyping as the above may unfold stuff in the actuals
+    for requires in sub_sig.requires() {
+        infcx.check_pred(requires, ConstrReason::SubtypeReq);
     }
 
     // 4. Plug in the EVAR solution / replace evars
@@ -325,15 +311,12 @@ fn check_fn_subtyping(
         .output()
         .replace_bound_refts_with(|sort, mode, _| infcx.fresh_infer_var(sort, mode));
     infcx
-        .subtyping(&output.ret, &super_output.ret, ConstrReason::Ret)
+        .subtyping(&output.ret, &super_output.ret, ConstrReason::SubtypeOut)
         .with_span(span)?;
 
-    // println!("TRACE: check_fn_subtyping (0): {env:?}");
-    // // 6. Update state with Output "ensures"
+    // 6. Update state with Output "ensures" and check super ensures
     update_ensures(&mut infcx, &mut env, &output, overflow_checking)?;
-    // println!("TRACE: check_fn_subtyping (1): {env:?}");
-    // // 7. Check ensures predicates
-    check_ensures(&mut infcx, &mut env, &super_output, span)?;
+    check_ensures(&mut infcx, &mut env, &super_output, ConstrReason::SubtypeEns, span)?;
 
     let evars_sol = infcx.pop_scope().with_span(span)?;
     infcx.replace_evars(&evars_sol);
@@ -360,13 +343,18 @@ fn update_ensures(
     Ok(())
 }
 
-fn check_ensures(at: &mut InferCtxtAt, env: &mut TypeEnv, output: &FnOutput, span: Span) -> Result {
+fn check_ensures(
+    at: &mut InferCtxtAt,
+    env: &mut TypeEnv,
+    output: &FnOutput,
+    reason: ConstrReason,
+    span: Span,
+) -> Result {
     for constraint in &output.ensures {
         match constraint {
             Ensures::Type(path, ty) => {
                 let actual_ty = env.get(path);
-                at.subtyping(&actual_ty, ty, ConstrReason::Ret)
-                    .with_span(span)?;
+                at.subtyping(&actual_ty, ty, reason).with_span(span)?;
             }
             Ensures::Pred(e) => {
                 at.check_pred(e, ConstrReason::Ret);
@@ -782,7 +770,7 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             .subtyping(&ret_place_ty, &output.ret, ConstrReason::Ret)
             .with_span(span)?;
 
-        check_ensures(&mut at, env, &output, span)?;
+        check_ensures(&mut at, env, &output, ConstrReason::Ret, span)?;
 
         let evars_sol = infcx.pop_scope().with_span(span)?;
         infcx.replace_evars(&evars_sol);

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -4,7 +4,7 @@ use flux_common::{bug, dbg, index::IndexVec, iter::IterExt, tracked_span_bug};
 use flux_config as config;
 use flux_infer::{
     fixpoint_encoding::{self, KVarGen},
-    infer::{ConstrReason, InferCtxt, InferCtxtAt, InferCtxtRoot},
+    infer::{ConstrReason, InferCtxt, InferCtxtAt, InferCtxtRoot, SubtypeReason},
     refine_tree::{AssumeInvariants, RefineTree, Snapshot},
 };
 use flux_middle::{
@@ -287,13 +287,15 @@ fn check_fn_subtyping(
         infcx.assume_pred(requires);
     }
     for (actual, formal) in iter::zip(actuals, sub_sig.inputs()) {
+        let reason = ConstrReason::Subtype(SubtypeReason::Input);
         infcx
-            .fun_arg_subtyping(&mut env, &actual, formal, ConstrReason::SubtypeIn)
+            .fun_arg_subtyping(&mut env, &actual, formal, reason)
             .with_span(span)?;
     }
     // we check the requires AFTER the actual-formal subtyping as the above may unfold stuff in the actuals
     for requires in sub_sig.requires() {
-        infcx.check_pred(requires, ConstrReason::SubtypeReq);
+        let reason = ConstrReason::Subtype(SubtypeReason::Requires);
+        infcx.check_pred(requires, reason);
     }
 
     // 4. Plug in the EVAR solution / replace evars
@@ -310,14 +312,21 @@ fn check_fn_subtyping(
     let super_output = super_sig
         .output()
         .replace_bound_refts_with(|sort, mode, _| infcx.fresh_infer_var(sort, mode));
+    let reason = ConstrReason::Subtype(SubtypeReason::Output);
     infcx
-        .subtyping(&output.ret, &super_output.ret, ConstrReason::SubtypeOut)
+        .subtyping(&output.ret, &super_output.ret, reason)
         .with_span(span)?;
 
     // 6. Update state with Output "ensures" and check super ensures
     update_ensures(&mut infcx, &mut env, &output, overflow_checking)?;
     fold_local_ptrs(&mut infcx, &mut env, span)?;
-    check_ensures(&mut infcx, &mut env, &super_output, ConstrReason::SubtypeEns, span)?;
+    check_ensures(
+        &mut infcx,
+        &mut env,
+        &super_output,
+        ConstrReason::Subtype(SubtypeReason::Ensures),
+        span,
+    )?;
 
     let evars_sol = infcx.pop_scope().with_span(span)?;
     infcx.replace_evars(&evars_sol);

--- a/crates/flux-refineck/src/checker.rs
+++ b/crates/flux-refineck/src/checker.rs
@@ -318,7 +318,7 @@ fn check_fn_subtyping(
         .with_span(span)?;
 
     // 6. Update state with Output "ensures" and check super ensures
-    update_ensures(&mut infcx, &mut env, &output, overflow_checking)?;
+    update_ensures(&mut infcx, &mut env, &output, overflow_checking);
     fold_local_ptrs(&mut infcx, &mut env, span)?;
     check_ensures(
         &mut infcx,
@@ -339,7 +339,7 @@ fn update_ensures(
     env: &mut TypeEnv,
     output: &FnOutput,
     overflow_checking: bool,
-) -> Result {
+) {
     for ensure in &output.ensures {
         match ensure {
             Ensures::Type(path, updated_ty) => {
@@ -350,7 +350,6 @@ fn update_ensures(
             Ensures::Pred(e) => infcx.assume_pred(e),
         }
     }
-    Ok(())
 }
 
 fn check_ensures(
@@ -867,7 +866,7 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             .replace_evars(&evars_sol)
             .replace_bound_refts_with(|sort, _, _| infcx.define_vars(sort));
 
-        update_ensures(infcx, env, &output, self.check_overflow())?;
+        update_ensures(infcx, env, &output, self.check_overflow());
 
         fold_local_ptrs(infcx, env, span)?;
 

--- a/crates/flux-refineck/src/lib.rs
+++ b/crates/flux-refineck/src/lib.rs
@@ -207,9 +207,13 @@ fn report_errors(genv: GlobalEnv, errors: Vec<Tag>) -> Result<(), ErrorGuarantee
     for err in errors {
         let span = err.src_span;
         e = Some(match err.reason {
-            ConstrReason::Call => call_error(genv, span, err.dst_span),
+            ConstrReason::Call | ConstrReason::SubtypeIn | ConstrReason::SubtypeReq => {
+                call_error(genv, span, err.dst_span)
+            }
             ConstrReason::Assign => genv.sess().emit_err(errors::AssignError { span }),
-            ConstrReason::Ret => ret_error(genv, span, err.dst_span),
+            ConstrReason::Ret | ConstrReason::SubtypeOut | ConstrReason::SubtypeEns => {
+                ret_error(genv, span, err.dst_span)
+            }
             ConstrReason::Div => genv.sess().emit_err(errors::DivError { span }),
             ConstrReason::Rem => genv.sess().emit_err(errors::RemError { span }),
             ConstrReason::Goto(_) => genv.sess().emit_err(errors::GotoError { span }),

--- a/crates/flux-refineck/src/lib.rs
+++ b/crates/flux-refineck/src/lib.rs
@@ -37,7 +37,7 @@ use flux_common::{cache::QueryCache, dbg, result::ResultExt as _};
 use flux_config as config;
 use flux_infer::{
     fixpoint_encoding::{FixpointCtxt, KVarGen},
-    infer::{ConstrReason, Tag},
+    infer::{ConstrReason, SubtypeReason, Tag},
     refine_tree::RefineTree,
 };
 use flux_macros::fluent_messages;
@@ -207,13 +207,15 @@ fn report_errors(genv: GlobalEnv, errors: Vec<Tag>) -> Result<(), ErrorGuarantee
     for err in errors {
         let span = err.src_span;
         e = Some(match err.reason {
-            ConstrReason::Call | ConstrReason::SubtypeIn | ConstrReason::SubtypeReq => {
+            ConstrReason::Call
+            | ConstrReason::Subtype(SubtypeReason::Input)
+            | ConstrReason::Subtype(SubtypeReason::Requires) => {
                 call_error(genv, span, err.dst_span)
             }
             ConstrReason::Assign => genv.sess().emit_err(errors::AssignError { span }),
-            ConstrReason::Ret | ConstrReason::SubtypeOut | ConstrReason::SubtypeEns => {
-                ret_error(genv, span, err.dst_span)
-            }
+            ConstrReason::Ret
+            | ConstrReason::Subtype(SubtypeReason::Output)
+            | ConstrReason::Subtype(SubtypeReason::Ensures) => ret_error(genv, span, err.dst_span),
             ConstrReason::Div => genv.sess().emit_err(errors::DivError { span }),
             ConstrReason::Rem => genv.sess().emit_err(errors::RemError { span }),
             ConstrReason::Goto(_) => genv.sess().emit_err(errors::GotoError { span }),

--- a/crates/flux-refineck/src/lib.rs
+++ b/crates/flux-refineck/src/lib.rs
@@ -169,7 +169,8 @@ pub fn check_fn(
 
         // PHASE 4: subtyping check for trait-method implementations
         if let Some((refine_tree, kvars)) =
-            trait_impl_subtyping(genv, local_id, span).map_err(|err| err.emit_err(&genv, def_id))?
+            trait_impl_subtyping(genv, local_id, config.check_overflow, span)
+                .map_err(|err| err.emit_err(&genv, def_id))?
         {
             tracing::info!("check_fn::refine-subtyping");
             let errors =

--- a/crates/flux-refineck/src/type_env.rs
+++ b/crates/flux-refineck/src/type_env.rs
@@ -28,6 +28,7 @@ use flux_rustc_bridge::{
     ty,
 };
 use itertools::{izip, Itertools};
+use rustc_index::IndexSlice;
 use rustc_middle::{mir::RETURN_PLACE, ty::TyCtxt};
 use rustc_type_ir::BoundVar;
 
@@ -84,6 +85,25 @@ impl<'a> TypeEnv<'a> {
 
         env.alloc(RETURN_PLACE);
         env
+    }
+
+    pub fn empty() -> TypeEnv<'a> {
+        TypeEnv { bindings: PlacesTree::default(), local_decls: IndexSlice::empty() }
+        // for requires in fn_sig.requires() {
+        //     infcx.assume_pred(requires);
+        // }
+
+        // for ty in fn_sig.inputs() {
+        //     let ty = infcx.unpack(ty);
+        //     infcx.assume_invariants(&ty, check_overflow);
+        //     if let TyKind::StrgRef(_, path, ty) = ty.kind() {
+        //         let loc = path.to_loc().unwrap();
+        //         env.bindings.insert(loc, LocKind::Universal, ty.clone());
+        //     }
+        // }
+
+        // env.alloc(RETURN_PLACE);
+        // env
     }
 
     fn alloc_with_ty(&mut self, local: Local, ty: Ty) {
@@ -329,6 +349,21 @@ impl<'a> TypeEnv<'a> {
         Ok(loc)
     }
 
+    pub(crate) fn unfold_strg_ref(
+        &mut self,
+        infcx: &mut InferCtxt,
+        path: &Path,
+        ty: &Ty,
+    ) -> InferResult<Loc> {
+        if let Some(loc) = path.to_loc() {
+            let ty = infcx.unpack(ty);
+            self.bindings.insert(loc, LocKind::Universal, ty);
+            Ok(loc)
+        } else {
+            bug!("unfold_strg_ref: unexpected path {path:?}")
+        }
+    }
+
     pub(crate) fn unfold(
         &mut self,
         infcx: &mut InferCtxt,
@@ -379,6 +414,10 @@ impl flux_infer::infer::LocEnv for TypeEnv<'_> {
 
     fn get(&self, path: &Path) -> Ty {
         self.get(path)
+    }
+
+    fn unfold_strg_ref(&mut self, infcx: &mut InferCtxt, path: &Path, ty: &Ty) -> InferResult<Loc> {
+        self.unfold_strg_ref(infcx, path, ty)
     }
 }
 

--- a/crates/flux-refineck/src/type_env.rs
+++ b/crates/flux-refineck/src/type_env.rs
@@ -16,9 +16,9 @@ use flux_middle::{
         evars::EVarSol,
         fold::{FallibleTypeFolder, TypeFoldable, TypeVisitable, TypeVisitor},
         region_matching::{rty_match_regions, ty_match_regions},
-        BaseTy, Binder, BoundReftKind, Expr, ExprKind, FnSig, GenericArg, HoleKind, Lambda, List,
-        Loc, Mutability, Path, PtrKind, Region, SortCtor, SubsetTy, Ty, TyKind, VariantIdx,
-        INNERMOST,
+        BaseTy, Binder, BoundReftKind, Ensures, Expr, ExprKind, FnOutput, FnSig, GenericArg,
+        HoleKind, Lambda, List, Loc, Mutability, Path, PtrKind, Region, SortCtor, SubsetTy, Ty,
+        TyKind, VariantIdx, INNERMOST,
     },
     PlaceExt as _,
 };
@@ -380,6 +380,44 @@ impl<'a> TypeEnv<'a> {
     pub fn replace_evars(&mut self, evars: &EVarSol) {
         self.bindings
             .fmap_mut(|binding| binding.replace_evars(evars));
+    }
+
+    pub(crate) fn update_ensures(
+        &mut self,
+        infcx: &mut InferCtxt,
+        output: &FnOutput,
+        overflow_checking: bool,
+    ) {
+        for ensure in &output.ensures {
+            match ensure {
+                Ensures::Type(path, updated_ty) => {
+                    let updated_ty = infcx.unpack(updated_ty);
+                    infcx.assume_invariants(&updated_ty, overflow_checking);
+                    self.update_path(path, updated_ty);
+                }
+                Ensures::Pred(e) => infcx.assume_pred(e),
+            }
+        }
+    }
+
+    pub(crate) fn check_ensures(
+        &mut self,
+        at: &mut InferCtxtAt,
+        output: &FnOutput,
+        reason: ConstrReason,
+    ) -> InferResult {
+        for constraint in &output.ensures {
+            match constraint {
+                Ensures::Type(path, ty) => {
+                    let actual_ty = self.get(path);
+                    at.subtyping(&actual_ty, ty, reason)?;
+                }
+                Ensures::Pred(e) => {
+                    at.check_pred(e, ConstrReason::Ret);
+                }
+            }
+        }
+        Ok(())
     }
 }
 

--- a/tests/tests/neg/surface/impl01.rs
+++ b/tests/tests/neg/surface/impl01.rs
@@ -1,0 +1,14 @@
+pub trait Mono {
+    #[flux::sig(fn (zing: &strg i32[@n]) ensures zing: i32{v:n < v})]
+    fn foo(z: &mut i32);
+}
+
+pub struct Horse;
+
+impl Mono for Horse {
+    #[flux::sig(fn (z: &strg i32[@n]) ensures z: i32{v:v < n})]
+    fn foo(z: &mut i32) {
+        //~^ ERROR refinement type
+        *z -= 1;
+    }
+}

--- a/tests/tests/neg/surface/impl02.rs
+++ b/tests/tests/neg/surface/impl02.rs
@@ -1,0 +1,29 @@
+pub trait Mono {
+    #[flux::sig(fn (zing: &strg i32[@n])
+                  requires 0 < n
+                  ensures zing: i32{v:n < v})]
+    fn foo(z: &mut i32);
+}
+
+pub struct Tiger;
+
+impl Mono for Tiger {
+    #[flux::sig(fn (pig: &strg i32[@m])
+                  requires 100 < m
+                  ensures pig: i32{v:m < v})]
+    fn foo(pig: &mut i32) {
+        //~^ ERROR refinement type
+        *pig += 1;
+    }
+}
+
+pub struct Snake;
+
+impl Mono for Snake {
+    #[flux::sig(fn (pig: &strg {i32[@m] | 100 < m})
+                ensures pig: i32[m+1])]
+    fn foo(pig: &mut i32) {
+        //~^ ERROR refinement type
+        *pig += 1;
+    }
+}

--- a/tests/tests/neg/surface/impl03.rs
+++ b/tests/tests/neg/surface/impl03.rs
@@ -1,0 +1,14 @@
+pub trait Mono {
+    #[flux::sig(fn (zing: &mut i32{v: 0 <= v}))]
+    fn foo(z: &mut i32);
+}
+
+pub struct Snake;
+
+impl Mono for Snake {
+    #[flux::sig(fn (hog: &strg i32[@m]) ensures hog: i32[m-1])]
+    fn foo(hog: &mut i32) {
+        //~^ ERROR: type invariant may not hold
+        *hog -= 1;
+    }
+}

--- a/tests/tests/pos/surface/impl01.rs
+++ b/tests/tests/pos/surface/impl01.rs
@@ -1,0 +1,22 @@
+pub trait Mono {
+    #[flux::sig(fn (zing: &strg i32[@n]) ensures zing: i32{v:n < v})]
+    fn foo(z: &mut i32);
+}
+
+pub struct Tiger;
+
+impl Mono for Tiger {
+    #[flux::sig(fn (pig: &strg i32[@m]) ensures pig: i32{v:m < v})]
+    fn foo(pig: &mut i32) {
+        *pig += 1;
+    }
+}
+
+pub struct Snake;
+
+impl Mono for Snake {
+    #[flux::sig(fn (pig: &strg i32[@m]) ensures pig: i32[m+1])]
+    fn foo(pig: &mut i32) {
+        *pig += 1;
+    }
+}

--- a/tests/tests/pos/surface/impl02.rs
+++ b/tests/tests/pos/surface/impl02.rs
@@ -1,0 +1,27 @@
+pub trait Mono {
+    #[flux::sig(fn (zing: &strg i32[@n])
+                  requires 100 < n
+                  ensures zing: i32{v:n < v})]
+    fn foo(z: &mut i32);
+}
+
+pub struct Tiger;
+
+impl Mono for Tiger {
+    #[flux::sig(fn (pig: &strg i32[@m])
+                  requires 0 < m
+                  ensures pig: i32{v:m < v})]
+    fn foo(pig: &mut i32) {
+        *pig += 1;
+    }
+}
+
+pub struct Snake;
+
+impl Mono for Snake {
+    #[flux::sig(fn (pig: &strg {i32[@m] | 0 < m})
+                ensures pig: i32[m+1])]
+    fn foo(pig: &mut i32) {
+        *pig += 1;
+    }
+}

--- a/tests/tests/pos/surface/impl03.rs
+++ b/tests/tests/pos/surface/impl03.rs
@@ -1,0 +1,13 @@
+pub trait Mono {
+    #[flux::sig(fn (zing: &mut i32{v: 0 <= v}))]
+    fn foo(z: &mut i32);
+}
+
+pub struct Tiger;
+
+impl Mono for Tiger {
+    #[flux::sig(fn (pig: &strg i32[@m]) ensures pig: i32[m+1])]
+    fn foo(pig: &mut i32) {
+        *pig += 1;
+    }
+}

